File: week12/2/Prob3.cpp
#include<vector>

using namespace std;

class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int count = 0;
        int m = board.size();
        int n = board[0].size();
        
        for(int i = 0; i < m; ++i){
            for(int j = 0; j < n; ++j){
                if(board[i][j] == 'X'){
                    // 전함의 시작 부분인지 체크
                    if(i > 0 && board[i - 1][j] == 'X'){
                        // 위에 전함이 존재하면 이는 새로운 전함이 아님
                        continue;
                    }
                    if(j > 0 && board[i][j - 1] == 'X'){
                        // 왼쪽에 전함이 존재하면 이는 새로운 전함이 아님
                        continue;
                    }
                    // 새로운 전함으로 판단, count 증가
                    count++;
                }
            }
        }
        
        return count;
    }
};

File: week12/2/Prob3.py
from typing import List

class Solution:
    def countBattleships(self, board: List[List[str]]) -> int:
        if not board or not board[0]:
            return 0
        
        rows, cols = len(board), len(board[0])
        count = 0
        
        for r in range(rows):
            for c in range(cols):
                # battleship의 top left에서 count를 수행, 나머지 위치에서는 X가 나오더라도 count하지 않음
                if board[r][c] == 'X':
                    if (r == 0 or board[r - 1][c] != 'X') and (c == 0 or board[r][c - 1] != 'X'):
                        count += 1
        
        return count